primitives/trie/src/cache/mod.rs

```diff
@@ -36,43 +36,285 @@
+/// cache can replace in one go.
+///
+/// We don't want a single local cache instance to have the ability to replace
+/// everything in the shared cache.
+const SHARED_NODE_CACHE_MAX_REPLACE_PERCENT: usize = 33;
+/// Same as [`SHARED_NODE_CACHE_MAX_REPLACE_PERCENT`].
+const SHARED_VALUE_CACHE_MAX_REPLACE_PERCENT: usize = 33;
+
+/// The maximum inline capacity of the local cache, in bytes.
+///
+/// This is just an upper limit; since the maps are resized in powers of two
+/// their actual size will most likely not exactly match this.
+const LOCAL_NODE_CACHE_MAX_INLINE_SIZE: usize = 512 * 1024;
+/// Same as [`LOCAL_NODE_CACHE_MAX_INLINE_SIZE`].
+const LOCAL_VALUE_CACHE_MAX_INLINE_SIZE: usize = 512 * 1024;
+
+/// The maximum size of the memory allocated on the heap by the local cache, in bytes.
+const LOCAL_NODE_CACHE_MAX_HEAP_SIZE: usize = 2 * 1024 * 1024;
+/// Same as [`LOCAL_NODE_CACHE_MAX_HEAP_SIZE`].
+const LOCAL_VALUE_CACHE_MAX_HEAP_SIZE: usize = 4 * 1024 * 1024;
+
+/// The size of the shared cache.
 #[derive(Debug, Clone, Copy)]
-pub enum CacheSize {
-	/// Do not limit the cache size.
-	Unlimited,
-	/// Let the cache in maximum use the given amount of bytes.
-	Maximum(usize),
-}
+pub struct CacheSize(usize);
 
 impl CacheSize {
-	/// Returns `true` if the `current_size` exceeds the allowed size.
-	fn exceeds(&self, current_size: usize) -> bool {
-		match self {
-			Self::Unlimited => false,
-			Self::Maximum(max) => *max < current_size,
+	/// An unlimited cache size.
+	pub const fn unlimited() -> Self {
+		CacheSize(usize::MAX)
+	}
+
+	/// A cache size `bytes` big.
+	pub const fn new(bytes: usize) -> Self {
+		CacheSize(bytes)
+	}
+}
+
+/// A limiter for the local node cache. This makes sure the local cache doesn't grow too big.
+#[derive(Default)]
+pub struct LocalNodeCacheLimiter {
+	/// The current size (in bytes) of data allocated by this cache on the heap.
+	///
+	/// This doesn't include the size of the map itself.
+	current_heap_size: usize,
+}
+
+impl<H> schnellru::Limiter<H, NodeCached<H>> for LocalNodeCacheLimiter
+where
+	H: AsRef<[u8]> + std::fmt::Debug,
+{
+	type KeyToInsert<'a> = H;
+	type LinkType = u32;
+
+	#[inline]
+	fn is_over_the_limit(&self, length: usize) -> bool {
+		// Only enforce the limit if there's more than one element to make sure
+		// we can always add a new element to the cache.
+		if length <= 1 {
+			return false
 		}
+
+		self.current_heap_size > LOCAL_NODE_CACHE_MAX_HEAP_SIZE
+	}
+
+	#[inline]
+	fn on_insert<'a>(
+		&mut self,
+		_length: usize,
+		key: H,
+		cached_node: NodeCached<H>,
+	) -> Option<(H, NodeCached<H>)> {
+		self.current_heap_size += cached_node.heap_size();
+		Some((key, cached_node))
+	}
+
+	#[inline]
+	fn on_replace(
+		&mut self,
+		_length: usize,
+		_old_key: &mut H,
+		_new_key: H,
+		old_node: &mut NodeCached<H>,
+		new_node: &mut NodeCached<H>,
+	) -> bool {
+		debug_assert_eq!(_old_key.as_ref().len(), _new_key.as_ref().len());
```

Comment:
This should never trigger; just a sanity check/docs-like assert to make sure the impls on the key are not broken, and it's partially also a vestigial remain from an earlier version of the limiter where I tracked the key sizes manually.

I guess it might be kind of useless and I could remove it if you want? (Technically it's not incorrect though.)